File: Go/is_subtree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isSameTree(s *TreeNode, t *TreeNode) bool {
    if s == nil && t == nil {return true}
    if s == nil || t == nil {return false}

    return s.Val == t.Val && isSameTree(s.Left, t.Left) && isSameTree(s.Right, t.Right)
}

func isSubtree(s *TreeNode, t *TreeNode) bool {
    if t == nil {return true}
    if s == nil {return false}

    return isSameTree(s, t) || isSubtree(s.Left, t) || isSubtree(s.Right, t)
}

File: Java/IsSubTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean equals(TreeNode s, TreeNode t) {
        if(s == null && t == null) return true;
        if(s == null || t == null) return false; 

        return s.val == t.val && equals(s.left, t.left) && equals(s.right, t.right);
    }

    public boolean isSubtree(TreeNode s, TreeNode t) {
        if(t == null) return true;
        if(s == null) return false;

        return equals(s, t) || isSubtree(s.left, t) || isSubtree(s.right, t);
    }
}